<a href="https://colab.research.google.com/github/NataliaGon/kpi/blob/AI-cybersecurity/Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import os
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import kagglehub
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import make_pipeline
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [6]:
path = kagglehub.dataset_download("purusinghvi/email-spam-classification-dataset")

print("Path to dataset files:", path)
print("Files in dataset folder:", os.listdir(path))

Using Colab cache for faster access to the 'email-spam-classification-dataset' dataset.
Path to dataset files: /kaggle/input/email-spam-classification-dataset
Files in dataset folder: ['combined_data.csv']


In [7]:
df = pd.read_csv(os.path.join(path, "combined_data.csv"))

print("Size rows/columns:", df.shape)

print(df.isnull().sum())


Size rows/columns: (83448, 2)
label    0
text     0
dtype: int64


In [9]:
X_train, X_test, y_train, y_test = train_test_split(df['text'], df['label'], test_size=0.2, random_state=42)

In [10]:
# alpha=1.0 applies Laplace smoothing to avoid zero probabilities
model = make_pipeline(CountVectorizer(), MultinomialNB(alpha=1.0))

model.fit(X_train, y_train)

y_pred = model.predict(X_test)


In [12]:
print(" Accuracy:", accuracy_score(y_test, y_pred))
print("\n Classification Report:\n", classification_report(y_test, y_pred))
print("\n Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

 Accuracy: 0.9761533852606351

 Classification Report:
               precision    recall  f1-score   support

           0       0.96      0.99      0.98      7938
           1       0.99      0.97      0.98      8752

    accuracy                           0.98     16690
   macro avg       0.98      0.98      0.98     16690
weighted avg       0.98      0.98      0.98     16690


 Confusion Matrix:
 [[7831  107]
 [ 291 8461]]


# Висновки

Після розподілу датасету на навчальну та тестову вибірки (у співвідношенні 80/20) модель була навчена на навчальних даних та протестована на незалежній тестовій вибірці. Отримані результати свідчать про високу ефективність алгоритму:

Точність моделі (Accuracy) становить 0.976, тобто близько 97,6% повідомлень було класифіковано правильно.

Показники precision, recall та f1-score для обох класів (spam та ham) перебувають у межах 0.96–0.99, що свідчить про збалансовану роботу класифікатора як для виявлення спаму, так і для розпізнавання звичайних повідомлень.

Матриця плутанини (Confusion Matrix) показує, що модель допустила лише 107 помилкових класифікацій звичайних повідомлень як спам та 291 помилку навпаки, що є незначним показником при загальній кількості тестових прикладів 16 690.

Додатково було розглянуто проблему нульової ймовірності, яка виникає у класичній формулі Байєса у випадку, коли певна ознака (слово) відсутня в навчальній вибірці для конкретного класу. Для розв’язання цієї проблеми застосовано оцінку Лапласа (Laplace smoothing) через параметр alpha=1.0 у моделі MultinomialNB. Це забезпечило ненульові значення ймовірностей для всіх ознак, підвищивши стабільність та достовірність моделі.

Таким чином, проведене дослідження підтвердило, що наївний байєсівський класифікатор є простим, швидким і ефективним методом класифікації текстів, який демонструє високу точність при виявленні спаму навіть на великих наборах даних. Алгоритм може бути рекомендований для застосування у системах фільтрації електронної пошти, моніторингу соціальних мереж, аналізу відгуків користувачів та інших завданнях класифікації текстів.